hazelcast/src/test/java/com/hazelcast/map/ExpirationTimeTest.java

```diff
@@ -51,6 +51,48 @@
 
     private static final long ONE_MINUTE_IN_MILLIS = MINUTES.toMillis(1);
 
+    @Test
+    public void prefer_ttl_expiry_over_max_idle_when_ttl_is_smaller() {
+        IMap<Integer, Integer> map = createMap();
+        final long ttlSeconds = 5;
+        final long maxIdleSeconds = 10;
+
+        map.put(1, 1, ttlSeconds, SECONDS, maxIdleSeconds, SECONDS);
+
+        long expirationTimeAfterPut = getExpirationTime(map, 1);
+
+        // multiple access
+        for (int i = 0; i < 3; i++) {
+            sleepSeconds(1);
+            map.get(1);
+        }
+
+        long expirationTimeAfterGet = getExpirationTime(map, 1);
+
+        assertEquals(expirationTimeAfterGet, expirationTimeAfterPut);
+    }
+
+    @Test
+    public void prefer_max_idle_expiry_over_ttl_when_max_idle_is_smaller() {
+        IMap<Integer, Integer> map = createMap();
+        final long ttlSeconds = 12;
+        final long maxIdleSeconds = 10;
+
+        map.put(1, 1, ttlSeconds, SECONDS, maxIdleSeconds, SECONDS);
+
+        long expirationTimeAfterPut = getExpirationTime(map, 1);
+
+        // multiple access
+        for (int i = 0; i < 3; i++) {
+            sleepSeconds(1);
+            map.get(1);
+        }
+
+        long expirationTimeAfterGet = getExpirationTime(map, 1);
+
+        assertTrue(expirationTimeAfterGet > expirationTimeAfterPut);
```

Comment:
`lastUpdateTime(expirationTimeAfterPut) + ttl >= expirationTimeAfterGet` must also hold for this case. But, the current implementation doesn't satisfy this condition  in this `ttl > maxIdleTime` case. Expiry time keeps increasing after exceeding the threshold of `ttl + lastUpdateTime`.

Reply:
Yes expiry time keeps increasing. When both ttl and maxIdle are configured we don't take into account bigger one, next expiration time will be always calculated over `maxIdleTime`.

Reply:
I think we should also consider the ttl in this case.
Although I do not know the previous behavior and implementation details of IMDG in this regard (maybe trying to apply these two policies together can be very costly according to the current implementation), I think that if the user provides these two policies, we should apply both of them. Whichever of these policies can evict an entry before, that policy entry should be able to evict that entry. In the example above, normally ttl will be able to evict this entry, but since we apply only one policy, the entry will live. Nor does [The API specification of IMap](https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/map/IMap.java#L1268-L1284) say that only one of these policies will apply.
Otherwise, I don't think it makes sense to provide an API that combines these policies anyway. The user can consider which number is smaller and set the desired policy more explicitly from the API containing only ttl or only maxIdleTime.

Reply:
Please see  #19483 and #19480. Now it should work as expected.